cmd/in-place-restart-agent/main.go

```diff
@@ -0,0 +1,352 @@
+func (r *InPlaceRestartAgent) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("namespace", r.Namespace, "podName", r.PodName, "inPlaceRestartExitCode", r.InPlaceRestartExitCode, "workerCommand", r.WorkerCommand, "podInPlaceRestartAttempt", r.PodInPlaceRestartAttempt, "isBarrierActive", r.IsBarrierActive)
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.Info("reconciling")
+
+	// Get associated JobSet
+	var js jobset.JobSet
+	if err := r.Get(ctx, req.NamespacedName, &js); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	currentInPlaceRestartAttempt := js.Status.CurrentInPlaceRestartAttempt
+	previousInPlaceRestartAttempt := js.Status.PreviousInPlaceRestartAttempt
+	log = log.WithValues("currentInPlaceRestartAttempt", currentInPlaceRestartAttempt, "previousInPlaceRestartAttempt", previousInPlaceRestartAttempt)
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Handle start up
+	// The Pod in-place restart attempt is set only once per agent execution, instead of possibly setting it multiple times in different reconciliations
+	// This could technically be done in the main function by getting the associated JobSet and patching the Pod, but it is done in the reconcile function to reuse the manager client and its watch
+	// This is done to decrease the impact on the API server as much as possible
+	// The startup is indicated by the Pod in-place restart attempt being not being set yet
+	if r.PodInPlaceRestartAttempt == nil {
+		var newPodInPlaceRestartAttempt int32
+		// If the JobSet current in-place restart attempt has not been set yet, it means that the JobSet is starting for the first time
+		// So set the Pod in-place restart attempt to 0
+		if currentInPlaceRestartAttempt == nil {
+			newPodInPlaceRestartAttempt = 0
+			// Otherwise, it means that the JobSet is restarting
+			// So set the Pod in-place restart attempt to the JobSet current in-place restart attempt + 1
+			// This will make sure the agent keeps waiting at the barrier until the JobSet controller lifts it (i.e., update the JobSet current in-place restart attempt)
+		} else {
+			newPodInPlaceRestartAttempt = *currentInPlaceRestartAttempt + 1
+		}
+		err := r.patchPodInPlaceRestartAttempt(ctx, newPodInPlaceRestartAttempt)
+		return ctrl.Result{}, err
+	}
+
+	// Handle in-place restart
+	// If the Pod in-place restart attempt is less than or equal to the previous in-place restart attempt, it means that the JobSet controller has marked the Pod in-place restart attempt as outdated
+	// Which means that the Pod should be restarted to reach the new in-place restart attempt
+	// So exit with the in-place restart exit code
+	// This will trigger container restart since Pod.spec.restartPolicy = OnFailure
+	// Once RestartAllContainers is released upstream (k8s 1.35), this will trigger in-place container restart since Pod.spec.initContainers[].restartPolicyRules[].action = RestartAllContainers
```

Comment:
What dependency does this feature have on 1.35?
Will this be smart enough to work on 1.32, 1.33 and 1.34?

And what happens if RestartAllContainers feature gate is disabled in upstream?